grakel_replace/torch_wl_kernel.py

```diff
@@ -159,69 +161,83 @@ def __init__(self, n_iter: int = 5, *, normalize: bool = True) -> None:
         super().__init__()
         self.n_iter = n_iter
         self.normalize = normalize
-        self.device: torch.device = torch.device("cpu")
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
         self.label_dict: dict[tuple, int] = {}
         self.label_counter: int = 0
+        self.hash_module = torch.nn.Linear(2, 1, bias=False)
+        torch.nn.init.normal_(self.hash_module.weight)
 
-    def _wl_iteration(self, adj: torch.sparse.Tensor,
-                      labels: torch.Tensor) -> torch.Tensor:
-        """Perform one WL iteration to update node labels.
-        Concatenate own label with sorted neighbor labels.
-
-        Args:
-            adj: Sparse adjacency matrix
-            labels: Current node label tensor
-
-        Returns:
-            Updated node label tensor
-        """
-        # Ensure the adjacency matrix is in COO format
+    def _wl_iteration(self, adj: Tensor, labels: Tensor) -> Tensor:
+        """Perform one iteration of the WL algorithm to update node labels."""
         adj = adj.coalesce()
         indices = adj.indices()
-        adj.values()
-
-        # Get the neighbors for each node
         rows, cols = indices
-        neighbors = cols[rows]
-
-        # Create a list of combined labels for each node
-        combined_labels = []
-        for node in range(labels.size(0)):
-            node_neighbors = neighbors[rows == node]
-            node_neighbor_labels = labels[node_neighbors]
-            combined_label = (labels[node].item(), tuple(node_neighbor_labels.tolist()))
-            combined_labels.append(combined_label)
-
-        # Update the label dictionary and counter
-        new_labels = torch.empty_like(labels)
-        for i, label in enumerate(combined_labels):
-            if label not in self.label_dict:
-                self.label_dict[label] = self.label_counter
-                self.label_counter += 1
-            new_labels[i] = self.label_dict[label]
+        num_nodes = labels.size(0)
 
-        return new_labels
+        # Create a mask for each node's neighbors
+        neighbor_mask = torch.zeros((num_nodes, num_nodes), dtype=torch.bool,
+                                    device=self.device)
+        neighbor_mask[rows, cols] = True
```

Comment:
Pretty much a micro optization but if there's a `todense()` on the `indices`, you might be able to do something like `neighbor_mask = indices.todense().astype(torch.bool)`, assuming `0` indicates no neighbors in the `indices` array.